File: NeetCode/Backtracking/Combination_Sum/combination__sum.py
from typing import Optional, List


class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        results = []

        def dfs(index: int, curr: list, total: int):
            if total == target:
                results.append(curr.copy())
                return

            if total > target or index >= len(nums):
                return

            curr.append(nums[index])
            dfs(index, curr.copy(), total + nums[index])
            curr.pop()
            dfs(index + 1, curr, total)

        dfs(0, [], 0)
        return results


if __name__ == "__main__":
    sol = Solution()
    nums = [2, 5, 6, 9]
    target = 9

    print(sol.combinationSum(nums, target))
    print("Running Solution...")

File: NeetCode/Backtracking/Combination_Sum/Combination_Sum.cpp
// Combination_Sum.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class Solution
{
public:
    vector<vector<int>> combinationSum(vector<int> &nums, int target)
    {
        vector<int> curr;
        vector<vector<int>> results;
        dfs(nums, target, curr, results, 0, 0);
        return results;
    }

    void dfs(vector<int> &nums, int target, vector<int> &curr, vector<vector<int>> &results, int index, int total)
    {
        if (total == target)
        {
            results.push_back(curr);
            return;
        }

        if (total > target || index >= nums.size())
        {
            return;
        }

        curr.push_back(nums[index]);
        dfs(nums, target, curr, results, index, total + nums[index]);
        curr.pop_back();
        dfs(nums, target, curr, results, index + 1, total);
    }
};
